Code/Numerics/Conrec.h

```diff
@@ -0,0 +1,192 @@
+//
+//  Copyright (C) 2019 Greg Landrum
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <vector>
+#include <Geometry/point.h>
+
+namespace conrec {
+struct ConrecSegment {
+  RDGeom::Point2D p1;
+  RDGeom::Point2D p2;
+  double isoVal;
+  ConrecSegment(double x1, double y1, double x2, double y2, double isoVal)
+      : p1(x1, y1), p2(x2, y2), isoVal(isoVal){};
+  ConrecSegment(const RDGeom::Point2D &p1, const RDGeom::Point2D &p2,
+                double isoVal)
+      : p1(p1), p2(p2), isoVal(isoVal){};
+};
+// adapted from conrec.c by Paul Bourke:
+// http://paulbourke.net/papers/conrec/conrec.c
+/*
+   Derivation from the fortran version of CONREC by Paul Bourke
+   d               ! matrix of data to contour
+   ilb,iub,jlb,jub ! index bounds of data matrix
+   x               ! data matrix column coordinates
+   y               ! data matrix row coordinates
+   nc              ! number of contour levels
+   z               ! contour levels in increasing order
+*/
+inline void Contour(const double *d, size_t ilb, size_t iub, size_t jlb,
+                    size_t jub, const double *x, const double *y, size_t nc,
+                    double *z, std::vector<ConrecSegment> &res) {
+  PRECONDITION(d, "no data");
+  PRECONDITION(x, "no data");
+  PRECONDITION(y, "no data");
+  PRECONDITION(z, "no data");
+  PRECONDITION(nc > 0, "no contours");
+  PRECONDITION(iub > ilb, "bad bounds");
+  PRECONDITION(jub > jlb, "bad bounds");
+
+#define xsect(p1, p2) (h[p2] * xh[p1] - h[p1] * xh[p2]) / (h[p2] - h[p1])
+#define ysect(p1, p2) (h[p2] * yh[p1] - h[p1] * yh[p2]) / (h[p2] - h[p1])
+
+  int m1, m2, m3, case_value;
+  double dmin, dmax, x1 = 0, x2 = 0, y1 = 0, y2 = 0;
+  int i, j, m;
```

Comment:
So FORTRANY!

Reply:
yep
there was no way I was going to rewrite that code though.